File: daily/2025/1261_Find_Elements_in_a_Contaminated_Binary_Tree.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

TreeNode* create_tree(vector<int>& nodes){
    if (nodes.empty()) return nullptr;
    TreeNode* root = new TreeNode(nodes[0]);
    int n = nodes.size();
    int i = 1;
    queue<TreeNode*> que;
    que.push(root);
    while(!que.empty()){
        TreeNode* curr = que.front();
        que.pop();
        if (i < n && nodes[i] != -1){
            curr->left = new TreeNode(nodes[i]);
            que.push(curr->left);
        }
        i++;
        if (i < n && nodes[i] != -1){
            curr->right = new TreeNode(nodes[i]);
            que.push(curr->right);
        }
        i++;
    }
    return root;
}

class FindElements {
public:
    unordered_set<int> s;
    FindElements(TreeNode* root) {
        root->val = 0;
        recursion(root);
    }

    void recursion(TreeNode* root){
        if (!root) return;
        s.insert(root->val);
        if (root->left){
            root->left->val = root->val * 2 + 1;
            recursion(root->left);
        }
        if (root->right){
            root->right->val = root->val * 2 + 2;
            recursion(root->right);
        }
    }
    bool find(int target) {
        return s.find(target) != s.end();
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */

int main(){
    vector<int> nodes = {0,-1,0};
    vector<int> query = {1,2};
    // vector<int> nodes = {0,0,0,0,0};
    // vector<int> query = {1,3,5};
    // vector<int> nodes = {0,-1,0,0,-1,0};
    // vector<int> query = {2,3,4,5};
    TreeNode* root = create_tree(nodes);
    vector<int> ans;
    FindElements F(root);
    for (int q : query) ans.push_back(F.find(q));
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/1261_Find_Elements_in_a_Contaminated_Binary_Tree.py
import os
import sys
from typing import List, Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Tree:
    def __init__(self, nodes):
        if not nodes:
            self.root = None
            return
        self.root = TreeNode(nodes[0])
        n = len(nodes)
        i = 1
        que = deque([self.root])
        while que:
            curr = que.popleft()
            if i < n and nodes[i] != -1:
                curr.left = TreeNode(nodes[i])
                que.append(curr.left)
            i += 1
            if i < n and nodes[i] != -1:
                curr.right = TreeNode(nodes[i])
                que.append(curr.right)
            i += 1

class FindElements:
    def __init__(self, root: Optional[TreeNode]):
        self.s = set()
        root.val = 0
        self.recursion(root)
    
    def recursion(self, node: Optional[TreeNode]):
        if not node:
            return
        self.s.add(node.val)
        if node.left:
            node.left.val = node.val * 2 + 1
            self.recursion(node.left)
        if node.right:
            node.right.val = node.val * 2 + 2
            self.recursion(node.right)

    def find(self, target: int) -> bool:
        return target in self.s

# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nodes = list(map(int,lines[0].split('=')[1].strip()[1:-2].split(',')))
            query = list(map(int,lines[1].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((nodes, query))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    for nodes, query in dataset:
        result = []
        T = Tree(nodes)
        F = FindElements(T.root)
        for q in query:
            result.append(F.find(q))
        results.append(result)
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
